src/passes/OptimizeInstructions.cpp

```diff
@@ -1196,185 +1231,165 @@ struct OptimizeInstructions
+      return curr;
+    }
+    // (unsigned)x / -1   ==>   x == -1
+    // TODO: i64 as well if sign-extension is enabled
+    if (matches(curr, binary(DivUInt32, any(), ival(-1)))) {
+      curr->op = Abstract::getBinary(type, Abstract::Eq);
+      return curr;
+    }
+    // x * -1   ==>   0 - x
+    if (matches(curr, binary(Abstract::Mul, any(&left), ival(-1)))) {
+      right->value = Literal::makeSingleZero(type);
+      curr->op = Abstract::getBinary(type, Abstract::Sub);
+      curr->left = right;
+      curr->right = left;
+      return curr;
+    }
+    // (unsigned)x <= -1   ==>   1
+    if (matches(curr, binary(Abstract::LeU, pure(&left), ival(-1)))) {
+      right->value = Literal::makeFromInt32(1, Type::i32);
+      right->type = Type::i32;
+      return right;
+    }
+    {
+      // Wasm binary encoding uses signed LEBs, which slightly favor negative
       // numbers: -64 is more efficient than +64 etc., as well as other powers
-      // of two 7 bits etc. higher. we therefore prefer x - -64 over x + 64.
-      // in theory we could just prefer negative numbers over positive, but
-      // that can have bad effects on gzip compression (as it would mean more
-      // subtractions than the more common additions).
-      if (binary->op == Abstract::getBinary(type, Abstract::Add) ||
-          binary->op == Abstract::getBinary(type, Abstract::Sub)) {
-        auto value = constRight;
-        if (value == 0x40 || value == 0x2000 || value == 0x100000 ||
-            value == 0x8000000 || value == 0x400000000LL ||
-            value == 0x20000000000LL || value == 0x1000000000000LL ||
-            value == 0x80000000000000LL || value == 0x4000000000000000LL) {
-          right->value = right->value.neg();
-          if (binary->op == Abstract::getBinary(type, Abstract::Add)) {
-            binary->op = Abstract::getBinary(type, Abstract::Sub);
-          } else {
-            binary->op = Abstract::getBinary(type, Abstract::Add);
-          }
-          return binary;
+      // of two 7 bits etc. higher. we therefore prefer x - -64 over x + 64. in
+      // theory we could just prefer negative numbers over positive, but that
+      // can have bad effects on gzip compression (as it would mean more
+      // subtractions than the more common additions). TODO: Simplify this by
+      // adding an ival matcher than can bind int64_t vars.
+      int64_t value;
+      if ((matches(curr, binary(Abstract::Add, any(), ival(&value))) ||
+           matches(curr, binary(Abstract::Sub, any(), ival(&value)))) &&
+          (value == 0x40 || value == 0x2000 || value == 0x100000 ||
+           value == 0x8000000 || value == 0x400000000LL ||
+           value == 0x20000000000LL || value == 0x1000000000000LL ||
+           value == 0x80000000000000LL || value == 0x4000000000000000LL)) {
+        right->value = right->value.neg();
+        if (matches(curr, binary(Abstract::Add, any(), constant()))) {
```

Comment:
this match is kind of wasteful, as we've matched most of it before... but maybe the optimizer figures it out? clarity-wise it seems a wash, so seems ok as it is.

Reply:
I hope the optimizer figures this out, but I'm not sure. I'll do another performance check before landing this.